File: coin_toss.go
// coin_toss.go — Go версия

package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

var coinFrames = []string{
	`
   _____
  /     \
 |  (●)  |
 |  (●)  |
  \_____/`,
	`
   _____
  /     \
 |  ( )  |
 |  ( )  |
  \_____/`,
}

type CoinToss3D struct {
	heads int
	tails int
	total int
}

func (c *CoinToss3D) clearScreen() {
	cmd := exec.Command("clear")
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cls")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func (c *CoinToss3D) animateCoin(result string) {
	for i := 0; i < 6; i++ {
		c.clearScreen()
		fmt.Println("🪙 3D Coin Toss (Go)")
		fmt.Println("Бросок...")
		frame := coinFrames[i%2]
		fmt.Println(frame)
		time.Sleep(80 * time.Millisecond)
	}

	c.clearScreen()
	fmt.Println("🪙 3D Coin Toss (Go)")
	if result == "heads" {
		fmt.Println("\x1b[32m")
		fmt.Println("   _____")
		fmt.Println("  /     \\")
		fmt.Println(" |  (●)  |")
		fmt.Println(" |  (●)  |")
		fmt.Println("  \\_____/")
		fmt.Println("\n✅ Результат: ОРЁЛ!\x1b[0m")
		c.heads++
	} else {
		fmt.Println("\x1b[31m")
		fmt.Println("   _____")
		fmt.Println("  /     \\")
		fmt.Println(" |  ( )  |")
		fmt.Println(" |  ( )  |")
		fmt.Println("  \\_____/")
		fmt.Println("\n❌ Результат: РЕШКА!\x1b[0m")
		c.tails++
	}
	c.total++
	fmt.Printf("\n📊 Статистика: Орёл: %d, Решка: %d\n", c.heads, c.tails)
}

func (c *CoinToss3D) flip() {
	rand.Seed(time.Now().UnixNano())
	result := "heads"
	if rand.Intn(2) == 0 {
		result = "tails"
	}
	c.animateCoin(result)
}

func (c *CoinToss3D) run() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("🪙 3D Coin Toss (Go)")
	fmt.Println("Нажмите Enter, чтобы бросить монету (q для выхода)")

	for {
		fmt.Print("> ")
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(strings.ToLower(input))
		if input == "q" {
			fmt.Println("До свидания!")
			break
		}
		c.flip()
	}
}

func main() {
	game := &CoinToss3D{}
	game.run()
}

File: coin_toss.js
// coin_toss.js — JavaScript версия

const readline = require('readline');

const coinFrames = [
    `
   _____
  /     \\
 |  (●)  |
 |  (●)  |
  \\_____/`,
    `
   _____
  /     \\
 |  ( )  |
 |  ( )  |
  \\_____/`
];

class CoinToss3D {
    constructor() {
        this.heads = 0;
        this.tails = 0;
        this.total = 0;
    }

    clearScreen() {
        console.clear();
    }

    animateCoin(result) {
        for (let i = 0; i < 6; i++) {
            this.clearScreen();
            console.log('🪙 3D Coin Toss (JavaScript)');
            console.log('Бросок...');
            console.log(coinFrames[i % 2]);
            // Синхронная задержка (блокирующая)
            const wait = Date.now() + 80;
            while (Date.now() < wait) {}
        }

        this.clearScreen();
        console.log('🪙 3D Coin Toss (JavaScript)');
        if (result === 'heads') {
            console.log('\x1b[32m');
            console.log('   _____');
            console.log('  /     \\');
            console.log(' |  (●)  |');
            console.log(' |  (●)  |');
            console.log('  \\_____/');
            console.log('\n✅ Результат: ОРЁЛ!\x1b[0m');
            this.heads++;
        } else {
            console.log('\x1b[31m');
            console.log('   _____');
            console.log('  /     \\');
            console.log(' |  ( )  |');
            console.log(' |  ( )  |');
            console.log('  \\_____/');
            console.log('\n❌ Результат: РЕШКА!\x1b[0m');
            this.tails++;
        }
        this.total++;
        console.log(`\n📊 Статистика: Орёл: ${this.heads}, Решка: ${this.tails}`);
    }

    flip() {
        const result = Math.random() < 0.5 ? 'heads' : 'tails';
        this.animateCoin(result);
    }

    run() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        console.log('🪙 3D Coin Toss (JavaScript)');
        console.log('Нажмите Enter, чтобы бросить монету (q для выхода)');

        rl.on('line', (input) => {
            if (input.trim().toLowerCase() === 'q') {
                console.log('До свидания!');
                rl.close();
                return;
            }
            this.flip();
        });
    }
}

const game = new CoinToss3D();
game.run();
